runtime/pkg/duckdbreplicator/conn.go

```diff
@@ -0,0 +1,127 @@
+package duckdbreplicator
+
+import (
+	"context"
+
+	"github.com/jmoiron/sqlx"
+)
+
+// Conn represents a single database connection.
+// This is useful when running a chain of queries using a single write connection.
+type Conn interface {
+	// Connx returns the underlying sqlx.Conn.
+	Connx() *sqlx.Conn
+
+	// CreateTableAsSelect creates a new table by name from the results of the given SQL query.
+	CreateTableAsSelect(ctx context.Context, name string, sql string, opts *CreateTableOptions) error
+
+	// InsertTableAsSelect inserts the results of the given SQL query into the table.
+	InsertTableAsSelect(ctx context.Context, name string, sql string, opts *InsertTableOptions) error
+
+	// DropTable removes a table from the database.
+	DropTable(ctx context.Context, name string) error
+
+	// RenameTable renames a table in the database.
+	RenameTable(ctx context.Context, oldName, newName string) error
+
+	// AddTableColumn adds a column to the table.
+	AddTableColumn(ctx context.Context, tableName, columnName, typ string) error
+
+	// AlterTableColumn alters the type of a column in the table.
+	AlterTableColumn(ctx context.Context, tableName, columnName, newType string) error
+}
+
+type conn struct {
+	*sqlx.Conn
+
+	db *db
+}
+
+var _ Conn = (*conn)(nil)
+
+func (c *conn) Connx() *sqlx.Conn {
+	return c.Conn
+}
+
+func (c *conn) CreateTableAsSelect(ctx context.Context, name, sql string, opts *CreateTableOptions) error {
+	if opts == nil {
+		opts = &CreateTableOptions{}
+	}
+	return c.db.createTableAsSelect(ctx, c.Conn, func() error { return nil }, name, sql, opts)
+}
+
+// InsertTableAsSelect inserts the results of the given SQL query into the table.
+func (c *conn) InsertTableAsSelect(ctx context.Context, name, sql string, opts *InsertTableOptions) error {
+	if opts == nil {
+		opts = &InsertTableOptions{
+			Strategy: IncrementalStrategyAppend,
+		}
+	}
+	return c.db.insertTableAsSelect(ctx, c.Conn, func() error { return nil }, name, sql, opts)
+}
+
+// DropTable removes a table from the database.
+func (c *conn) DropTable(ctx context.Context, name string) error {
+	return c.db.dropTable(ctx, name)
+}
+
+// RenameTable renames a table in the database.
+func (c *conn) RenameTable(ctx context.Context, oldName, newName string) error {
+	return c.db.renameTable(ctx, oldName, newName)
+}
+
+// AddTableColumn adds a column to the table.
+func (c *conn) AddTableColumn(ctx context.Context, tableName, columnName, typ string) error {
+	return c.db.addTableColumn(ctx, c.Conn, func() error { return nil }, tableName, columnName, typ)
+}
+
+// AlterTableColumn alters the type of a column in the table.
+func (c *conn) AlterTableColumn(ctx context.Context, tableName, columnName, newType string) error {
+	return c.db.alterTableColumn(ctx, c.Conn, func() error { return nil }, tableName, columnName, newType)
+}
+
+type singledbConn struct {
+	*sqlx.Conn
+
+	db *singledb
+}
+
+var _ Conn = (*singledbConn)(nil)
```

Comment:
The problem is that each needs to call private functions on underlying DB which are not interfaces. We could create interface that accepts `conn` parameter which both of these DBs implement but it is also somewhat unnecessary given there will not be more than two implementation of `DB` interface and the duplicated code is just a few lines.

This `singledb` is for two use cases:
1. For unit tests with in-memory db. I honestly think we should use the same interface even for unit tests. It is not ideal that most unit tests hit a different code path so this use case can be removed.
2. For cases when user wants queries to run against an external DB especially an in-memory DB. I have seen few users asking for that, though not sure why. If we are okay with removing this use case then I think it should be fine to remove `singledb`.

Reply:
About 2, I'm fine removing that use case (reads from an external DB can still be addressed with an `init_sql` that attaches the external DB). It would be nice to have all code run through one implementation path.